packages/flutter/lib/src/material/checkbox.dart

```diff
@@ -809,7 +853,166 @@ class _CheckboxPainter extends ToggleablePainter {
+    });
+  }
+
+  @override
+  MaterialStateProperty<Color> get fillColor {
+    return MaterialStateProperty.resolveWith((Set<MaterialState> states) {
+      if (states.contains(MaterialState.disabled)) {
+        if (states.contains(MaterialState.selected)) {
+          return _activeColor;
+        }
+        return Colors.transparent;
+      }
+      if (states.contains(MaterialState.selected)) {
+        if (states.contains(MaterialState.error)) {
+          return _colors.error;
+        }
+        return _activeColor;
+      }
+      return Colors.transparent;
+    });
   }
+
+  @override
+  MaterialStateProperty<Color> get checkColor {
+    return MaterialStateProperty.resolveWith((Set<MaterialState> states) {
+      if (states.contains(MaterialState.disabled)) {
+        if (states.contains(MaterialState.selected)) {
+          return CupertinoColors.white;
+        }
+        return Colors.transparent; // No icons available when the checkbox is unselected.
+      }
+      if (states.contains(MaterialState.selected)) {
+        if (states.contains(MaterialState.error)) {
+          return _colors.onError;
+        }
+        return CupertinoColors.white;
+      }
+      return Colors.transparent; // No icons available when the checkbox is unselected.
+    });
+  }
+
+  @override
+  MaterialStateProperty<Color> get overlayColor {
+    // The relative values needed to transform a color to it's equivalent focus
+    // outline color.
+    return MaterialStateProperty.resolveWith((Set<MaterialState> states) {
+      if (states.contains(MaterialState.error)) {
+        if (states.contains(MaterialState.pressed)) {
+          return _effectiveFocusOverlayColor;
+        }
+        if (states.contains(MaterialState.hovered)) {
+          return _effectiveFocusOverlayColor;
+        }
+        if (states.contains(MaterialState.focused)) {
+          return _effectiveFocusOverlayColor;
+        }
+      }
+      if (states.contains(MaterialState.selected)) {
+        if (states.contains(MaterialState.pressed)) {
+          return _effectiveFocusOverlayColor;
+        }
+        if (states.contains(MaterialState.hovered)) {
+          return _effectiveFocusOverlayColor;
+        }
+        if (states.contains(MaterialState.focused)) {
+          return _effectiveFocusOverlayColor;
+        }
+        return Colors.transparent;
+      }
+      if (states.contains(MaterialState.pressed)) {
+        return _effectiveFocusOverlayColor;
+      }
+      if (states.contains(MaterialState.hovered)) {
+        return _effectiveFocusOverlayColor;
+      }
+      if (states.contains(MaterialState.focused)) {
+        return _effectiveFocusOverlayColor;
+      }
+      return Colors.transparent;
+    });
+  }
+
+  @override
+  double get splashRadius => 40.0 / 2;
+
+  @override
+  MaterialTapTargetSize get materialTapTargetSize => _theme.materialTapTargetSize;
+
+  @override
+  VisualDensity get visualDensity => _theme.visualDensity;
```

Comment:
I wonder if there could be any unintended side effects of the Cupertino checkbox using values like visualDensity. I assume on native iOS, the checkbox has no concept of visual density. So could a user end up getting a weirdly sized checkbox on iOS due to the visualDensity they set for Material, when they wanted the density to match native iOS?